gcs/constraints/circuit.cc

```diff
@@ -20,184 +21,182 @@ using std::move;
+    auto propagate_circuit_using_prevent(const vector<IntegerVariableID> & succ,
+                                         const ProofLine2DMap & lines_for_setting_pos,
+                                         State & state,
+                                         const IntegerVariableID & trigger_var,
+                                         const long & trigger_idx,
+                                         const ConstraintStateHandle & chain_start_idx,
+                                         const ConstraintStateHandle & chain_end_idx,
+                                         const ConstraintStateHandle & chain_length_idx,
+                                         bool& should_disable) -> Inference
     {
+        // propagate all-different first, to avoid infinite loops
         // Have to use check first
-        auto result = propagate_circuit_using_check(succ, lines_for_setting_pos, state);
+        auto result = propagate_non_gac_alldifferent(trigger_var, trigger_idx, succ, state);
         if (result == Inference::Contradiction) return Inference::Contradiction;
 
-        vector<bool> checked(succ.size(), false);
-
-        // Assume all different has already been propagated (do we know this for sure ???)
-        for (const auto & var : succ) {
-            if (state.has_single_value(var)) continue;
-
-            // In that case the possible start points for chains are the domains of the unfixed variables
-            state.for_each_value_while(var, [&](auto val) {
-                if (checked[val.raw_value]) return true; // Followed this chain already
-                checked[val.raw_value] = true;
-
-                auto current_val = state.optional_single_value(succ[val.raw_value]);
-                auto chain_length = 1;
-                auto next_var = var;
-                if (current_val == nullopt) return true;
-
-                stringstream proof_step;
-                if (state.maybe_proof()) {
-                    proof_step << "p ";
-                    proof_step << lines_for_setting_pos.at(make_pair(current_val.value(), val)) + 1 << " ";
-                }
-
-                optional<Integer> last_val;
-
-                // Follow this chain to the end
-                while (current_val != nullopt) {
-                    last_val = current_val;
-                    next_var = succ[last_val->raw_value];
-                    current_val = state.optional_single_value(next_var);
-                    chain_length++;
-                    if (state.maybe_proof() && current_val != nullopt) {
-                        proof_step << lines_for_setting_pos.at(make_pair(current_val.value(), last_val.value())) + 1
-                                   << " + ";
-                    }
-                }
+        auto& chain_start = any_cast<vector<long long> &>(state.get_constraint_state(chain_start_idx));
```

Comment:
I wonder whether these should be in a struct, rather than having three lots of constraint state? I would be inclined to say yes, partly because any_cast is slooooooow.